src/nations/nations.cpp

```diff
@@ -140,21 +141,344 @@ void restore_unsaved_values(sys::state& state) {
+			p_prev = coast_0;
+
+			auto ps = path.size();
+			auto effective_distance = 0.f;
+			auto worst_movement_cost = 0.f;
+
+			for(size_t i = 0; i < ps; i++) {
+				auto p_current = path[i];
+				auto adj = state.world.get_province_adjacency_by_province_pair(p_prev, p_current);
+				float distance = province::distance(state, adj);
+				float sum_mods =
+					state.world.province_get_modifier_values(p_current, sys::provincial_mod_offsets::movement_cost)
+					+ state.world.province_get_modifier_values(p_prev, sys::provincial_mod_offsets::movement_cost);
+				effective_distance += std::max(0.01f, distance * std::max(0.01f, (sum_mods * 2.f + 1.0f)));
+				if(sum_mods > worst_movement_cost)
+					worst_movement_cost = std::max(0.01f, sum_mods);
+
+				p_prev = p_current;
+			}
+			state.world.trade_route_set_sea_distance(route, effective_distance / speed);
+		} else {
+			state.world.trade_route_set_sea_distance(route, 99999.f);
+		}
+
+		if(state.world.trade_route_get_is_land_route(route)) {
+			std::vector<dcon::province_id> path{ };
+			auto speed = base_speed;
+			dcon::province_id p_prev{ };
+
+			auto market_0_center = state.world.state_instance_get_capital(sid_0);
+			auto market_1_center = state.world.state_instance_get_capital(sid_1);
+			path = province::make_unowned_path(state, market_0_center, market_1_center);
+			speed *= 0.2f;
+			p_prev = market_0_center;
+
+			auto ps = path.size();
+			auto effective_distance = 0.f;
+			auto worst_movement_cost = 0.f;
+			auto min_railroad_level = 0.f;
+
+			for(size_t i = 0; i < ps; i++) {
+				auto p_current = path[i];
+				auto adj = state.world.get_province_adjacency_by_province_pair(p_prev, p_current);
+				float distance = province::distance(state, adj);
+				float sum_mods =
+					state.world.province_get_modifier_values(p_current, sys::provincial_mod_offsets::movement_cost)
+					+ state.world.province_get_modifier_values(p_prev, sys::provincial_mod_offsets::movement_cost);
+				float local_effective_distance = distance * std::max(0.01f, sum_mods * 3.f);
+				auto railroad_origin = state.world.province_get_building_level(p_prev, uint8_t(economy::province_building_type::railroad));
+				auto railroad_target = state.world.province_get_building_level(p_current, uint8_t(economy::province_building_type::railroad));
+				if(railroad_origin > 0 && railroad_target > 0) {
+					local_effective_distance = local_effective_distance / 2.f;
+				}
+				local_effective_distance -= 0.03f * std::min(railroad_target, railroad_origin) * local_effective_distance;
+				effective_distance += std::max(0.01f, local_effective_distance);
+				if(sum_mods > worst_movement_cost)
+					worst_movement_cost = std::max(0.01f, sum_mods);
+
+				p_prev = p_current;
+			}
+			state.world.trade_route_set_land_distance(route, effective_distance / speed);
+		} else {
+			state.world.trade_route_set_land_distance(route, 99999.f);
+		}
+	});
+}
+
+void generate_sea_trade_routes(sys::state& state) {
```

Comment:
This could probably also be parallelized, although with more effort. Basically instead of creating the new routes inside the main loop it would instead stick the information about the new route to be created into a thread-safe collection of some sort as it goes, and then do just the route creation at the end as a single-threaded operation. Again, not sure if that work is necessary or appropriate, however.